Update grammar `functionDefinition` parser rule and rename `NumberUnit` lexer rule (#14066)
ethereum/solidity: Changelog.md

Bugfixes:

>  * Antlr Grammar: Fix of a discrepancy which allowed numbers followed by an identifier with no whitespace.

Reviewer: ```suggestion
 * Antlr Grammar: Fix of a discrepancy with the parser, which allowed numbers followed by an identifier with no whitespace.
```

>  * Antlr Grammar: Stricter rules for function definitions will no longer accept as valid free functions having some specifiers which are exclusive of contract functions.

Reviewer: ```suggestion
 * Antlr Grammar: Stricter rules for function definitions. The grammar will no longer accept as valid free functions having specifiers which are exclusive to contract functions.
```

>  * Antlr Grammar: Fix divergent behaviour which allowed octal numbers.

Reviewer: ```suggestion
 * Antlr Grammar: Fix discrepancy with the parser, which allowed octal numbers.
```

Reviewer 2: This should go one up in order to keep the changelog sorted.

Bugfixes:

> 

Author: Not sure if it should have its own section like:
```
Antlr Grammar Changes:
 * Lexer: Fix of a discrepancy which allowed numbers followed by an identifier with no whitespace
 * Lexer: Fix divergent behaviour which allowed octal numbers.
 * Parser: More strict for free functions and will no longer accept some specifiers which are exclusive of contract functions.
```

Reviewer: Let's follow the way it was done in older entries.
